File: neetCode150/gasStation.js
// Greedy
// Time: O(n * m)
// Space: O(n)
var canCompleteCircuit = function(gas, cost) {

    const recursion = (idx, s, c) => {
        if (c < 0) return -1;
        if (s === idx) return s;
        if (s === -1) s = idx;

        c += (gas[idx] - cost[idx])
        idx = idx+1 >= gas.length ? 0 : idx+1

        return recursion(idx, s, c)
    }

    for (let i = 0; i < gas.length; i++) {
        const result = recursion(i, -1, 0)
        if (result >= 0) return result;
    }
    return -1
};


console.log(canCompleteCircuit(gas = [2,3,4], cost = [3,4,3]))

File: spaced_rep_java_oct_2024/gasStation.java
package spaced_rep_java_oct_2024;
// gas = [1,2,3,4,5], cost = [3,4,5,1,2]
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int currIDx = 0;
        int currVal = gas[0] - cost[0];
        // get best idx starting point
        for (int i = 1; i < gas.length; i++) {
           int val = gas[i] - cost[i];
           if (val >= currVal) {
            currIDx = i;
            currVal = val;
           }
        }
System.out.println("maxI " + currIDx + " currVal: " + currVal);
        return -1;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        System.out.println(sol.canCompleteCircuit(new int[]{1,2,3,4,5}, new int[]{3,4,5,1,2}));
    }
}